### PR TITLE
Retry HTTP::Client requests once if io is closed

### PR DESCRIPTION
**Pull request by carlhoerberg**
Earlier only reads would be retried, but a very common scenario in the
wild is that the HTTP server resets the client connection after some
time. When the persistent HTTP::Client then did a request it would fail,
and it couldn't even be recovered because `close` would raise an
exception too, never setting `@io` to nil.

Without this patch `HTTP::Client` can't be used in any real-world
scenarios using HTTP keepalive, at least not without monkey-patching
`HTTP::Client#close` and catching and retrying your self, not very user
friendly. Also confusing when read errors are retried but not write
errors.

This change makes HTTP::Client::Request.from_io raise an IO::EOFError if the
header can't be read, instead of simply returning nil. That exception is
caught once, the `@io` closed (by ignoring IO::Error in close), and retries
the request once more, next time a proper exception is raised to the
user.

### CODE REVIEW

**Comment by straight-shoota** [comment]
It's not a good idea to raise an exception, when the exception is just rescued again. Raising is quite costly and the previous code was intended to avoid this significant overhead.

I'm sure can make the intended change to behaviour without changing the internal implementations from signaling failure by return value. Improved semantics and performance are not mutually exclusive.

The minimal change to make the specs work is the error handling in `#close` plus this:
```diff
   private def exec_internal(request)
-    response = exec_internal_single(request)
+    begin
+      response = exec_internal_single(request)
+    rescue IO::Error
+      response = nil
+    end
     return handle_response(response) if response
```

There's no need to raise exceptions for all errors.

Specs for the yielding overload of `exec_internal` are missing.

**Comment by carlhoerberg** [comment]
Ok, sure, but exceptions are raised *everywhere*, it's only in this very unique case where you've been able to write to the socket but then the server closes the connection before returning anything. If the socket is closed before you've written, or when you're trying to close a broken socket (why would that raise in any case, especially if we're so performance sensitive??), we still have to catch and retry.

File: src/http/client.cr [hunk 1]
```diff
@@ -625,23 +629,25 @@ class HTTP::Client
...
         if response
-          return handle_response(response) do
-            yield response
-          end
+          return handle_response(response) { yield response }
         end
       end
+    rescue IO::Error
+      # retry
+    end
```
Comment (straight-shoota) [inline src/http/client.cr:640]: This is a bit tricky: The `rescue` should not cover `yield`. Raising an exception in the block must not trigger a retry. Only exceptions raised in `send_request` should be rescued in order to try again.
I think for this it's probably best to move the call to `send_request` out of `exec_internal_single`. At this point, maybe the latter method could be removed entirely? Not sure.

Comment (carlhoerberg) [reply src/http/client.cr]: Good point, fixed, and added a spec for it

**Review by straight-shoota — APPROVED** [review]

**Review by beta-ziliani — APPROVED** [review]
Thanks!

**Comment by z64** [comment]
As a heads up, this looks related to https://github.com/crystal-lang/crystal/pull/10113 (approved, but not milestoned). Will generate conflicts by the looks of it.

**Comment by beta-ziliani** [comment]
Nice catch @z64 🙇 
This PR seems to mostly supersede #10113, so I think it's best to work with this one.

**Comment by z64** [comment]
@beta-ziliani Sounds good. Although perhaps someone should review, and post for the record, what to do about the specific `EPIPE` check that #10113 does, but this MR does not. 

Is it unecessary? Or does it require futher review? My use case makes heavy use of HTTP connection pooling & re-use, so robustness of `Client` is important for me. But it's not immeidately clear the impact of the `EPIPE` check would have.

**Comment by carlhoerberg** [comment]
This retries all IO::Errors, including but not limited to EPIPE so you should be good.

**Pull request merged**